src/core/ref.jl

```diff
@@ -34,3 +34,75 @@ function find_ref_buses(pm::GenericPowerModel, nw)
 end
+
+"Adds arcs for TPPM transformers; for dclines and branches this is done in PMs"
+function add_arcs_trans!(pm::GenericPowerModel)
+    if !haskey(ref(pm, pm.cnw), :trans)
+        # this might happen when parsing data from matlab format
+        # the OpenDSS parser always inserts a trans dict
+        ref(pm, pm.cnw)[:trans] = Dict{Int, Any}()
+    end
+    # dirty fix add arcs_from/to_trans and bus_arcs_trans
+    pm.ref[:nw][0][:arcs_from_trans] = [(i, trans["f_bus"], trans["t_bus"]) for (i,trans) in ref(pm, :trans)]
+    pm.ref[:nw][0][:arcs_to_trans] = [(i, trans["t_bus"], trans["f_bus"]) for (i,trans) in ref(pm, :trans)]
+    pm.ref[:nw][0][:arcs_trans] = [pm.ref[:nw][0][:arcs_from_trans]..., pm.ref[:nw][0][:arcs_to_trans]...]
+    pm.ref[:nw][0][:bus_arcs_trans] = Dict{Int64, Array{Any, 1}}()
+    for i in ids(pm, :bus)
+        pm.ref[:nw][0][:bus_arcs_trans][i] = [e for e in pm.ref[:nw][0][:arcs_trans] if e[2]==i]
+    end
+end
+
+function calc_tp_trans_Tvi(pm::GenericPowerModel, i::Int; nw=pm.cnw)
+    trans = ref(pm, nw, :trans,  i)
+    # transformation matrices
+    # Tv and Ti will be compositions of these
+    #Tminn = [diagm(0=>ones(Float64, 3)) -ones(Float64, 3)]  # substract neutral from other conductors
+    #Tnoz = [diagm(0=>ones(Float64, 3)) zeros(Float64, 3)]   # remove the neutral
+    Tbr = [0 0 1; 1 0 0; 0 1 0]                             # barrel roll
+    Tdelt  = [1 -1 0; 0 1 -1; -1 0 1]                       # delta transform
+    # connection transformers
+    perm = trans["conn"][1:3]
+    #TODO allow this finally or not?
+    @assert(!(perm in ["213", "321", "132"]))
+    polarity = trans["conn"][4]
+    dyz = trans["conn"][5]
+    grounded = length(trans["conn"])>5 && trans["conn"][6]=='n'
+    # Tw will contain transformations related to permutation and polarity
+    perm_to_trans = Dict(
+        "123"=>diagm(0=>ones(Float64, 3)),
+        "312"=>Tbr,
+        "231"=>Tbr*Tbr
+    )
+    Tw = perm_to_trans[perm]
+    Tw = (polarity=='+') ? Tw : -Tw
+    #Tw = diagm(0=>ones(Float64, 3))
+    vmult = 1.0 # compensate for change in LN
+    if dyz=='y'
+        Tv_fr = Tw
+        Tv_im = diagm(0=>ones(Float64, 3))
+        Ti_fr = Tw
+        Ti_im = diagm(0=>ones(Float64, 3))
+        # if !grounded
+        #     # if not grounded, phase currents should sum to zero
+        #     Ti_fr = [Ti_fr; ones(1,3)]
+        #     Ti_im = [Ti_im; zeros(1,3)]
+        # end
+    elseif dyz=='d'
+        Tv_fr = Tdelt*Tw
+        Tv_im = diagm(0=>ones(Float64, 3))
+        Ti_fr = Tw
+        Ti_im = Tdelt'
+        vmult = sqrt(3)
+    elseif dyz=='z'
+        #TODO zif-zag here
+    else
+        error(LOGGER, "Illegal connection specification $dyz.")
+    end
+    #Tv_fr = diagm(0=>ones(Float64, 3))*Tminn
+    Cv_to = trans["vnom_kv"][1]*vmult/trans["vnom_kv"][2]
+    # make equations dimensionless
+    bkv_fr = ref(pm, nw, :bus, trans["f_bus"], "base_kv")
+    bkv_to = ref(pm, nw, :bus, trans["t_bus"], "base_kv")
```

Comment:
**wye transformer**
This ensures that the equations are always correct, even when the voltage base across a transformer does not scale proportionally to nominal tap ratio.
If we assume that the voltage base is scaled as such, this can be left out.

**delta transformer**
There is a tricky subtlety for a delta transformer. 
Say that the nominal voltage at the primary is V1, and at the secondary is V2. The nominal tap ratio then is V1:V2=V1/V2
However, a delta winding boosts up the voltage by a factor of sqrt(3). To obtain the nominal voltage at the secondary winding, the physical tap ratio (which should go into the equations) is a factor sqrt(3) higher, V1.sqrt(3) : V2 
Schematically:
V1---(delta-to-wye)---V1.sqrt(3)---N:1---V2
This is expressed by vmult in the code above.